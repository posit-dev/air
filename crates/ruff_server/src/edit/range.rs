// +------------------------------------------------------------+
// | Code adopted from:                                         |
// | Repository: https://github.com/astral-sh/ruff.git          |
// | Commit: 5bc9d6d3aa694ab13f38dd5cf91b713fd3844380           |
// +------------------------------------------------------------+

use super::PositionEncoding;
use biome_text_size::{TextRange, TextSize};
use lsp_types as types;
use ruff_source_file::OneIndexed;
use ruff_source_file::{LineIndex, SourceLocation};

pub(crate) trait RangeExt {
    fn to_text_range(&self, text: &str, index: &LineIndex, encoding: PositionEncoding)
        -> TextRange;
}

pub(crate) trait ToRangeExt {
    fn to_range(&self, text: &str, index: &LineIndex, encoding: PositionEncoding) -> types::Range;
}

fn u32_index_to_usize(index: u32) -> usize {
    usize::try_from(index).expect("u32 fits in usize")
}

impl RangeExt for lsp_types::Range {
    fn to_text_range(
        &self,
        text: &str,
        index: &LineIndex,
        encoding: PositionEncoding,
    ) -> TextRange {
        let start_line = index.line_range(
            OneIndexed::from_zero_indexed(u32_index_to_usize(self.start.line)),
            text,
        );
        let end_line = index.line_range(
            OneIndexed::from_zero_indexed(u32_index_to_usize(self.end.line)),
            text,
        );

        let (start_column_offset, end_column_offset) = match encoding {
            PositionEncoding::UTF8 => (
                TextSize::from(self.start.character),
                TextSize::from(self.end.character),
            ),

            PositionEncoding::UTF16 => {
                // Fast path for ASCII only documents
                if index.is_ascii() {
                    (
                        TextSize::from(self.start.character),
                        TextSize::from(self.end.character),
                    )
                } else {
                    // UTF16 encodes characters either as one or two 16 bit words.
                    // The position in `range` is the 16-bit word offset from the start of the line (and not the character offset)
                    // UTF-16 with a text that may use variable-length characters.
                    (
                        utf8_column_offset(self.start.character, &text[start_line]),
                        utf8_column_offset(self.end.character, &text[end_line]),
                    )
                }
            }
            PositionEncoding::UTF32 => {
                // UTF-32 uses 4 bytes for each character. Meaning, the position in range is a character offset.
                return TextRange::new(
                    index.offset(
                        OneIndexed::from_zero_indexed(u32_index_to_usize(self.start.line)),
                        OneIndexed::from_zero_indexed(u32_index_to_usize(self.start.character)),
                        text,
                    ),
                    index.offset(
                        OneIndexed::from_zero_indexed(u32_index_to_usize(self.end.line)),
                        OneIndexed::from_zero_indexed(u32_index_to_usize(self.end.character)),
                        text,
                    ),
                );
            }
        };

        TextRange::new(
            start_line.start() + start_column_offset.clamp(TextSize::from(0), start_line.end()),
            end_line.start() + end_column_offset.clamp(TextSize::from(0), end_line.end()),
        )
    }
}

impl ToRangeExt for TextRange {
    fn to_range(&self, text: &str, index: &LineIndex, encoding: PositionEncoding) -> types::Range {
        types::Range {
            start: source_location_to_position(&offset_to_source_location(
                self.start(),
                text,
                index,
                encoding,
            )),
            end: source_location_to_position(&offset_to_source_location(
                self.end(),
                text,
                index,
                encoding,
            )),
        }
    }
}

/// Converts a UTF-16 code unit offset for a given line into a UTF-8 column number.
fn utf8_column_offset(utf16_code_unit_offset: u32, line: &str) -> TextSize {
    let mut utf8_code_unit_offset = TextSize::from(0);

    let mut i = 0u32;

    for c in line.chars() {
        if i >= utf16_code_unit_offset {
            break;
        }

        // Count characters encoded as two 16 bit words as 2 characters.
        {
            utf8_code_unit_offset +=
                TextSize::from(u32::try_from(c.len_utf8()).expect("utf8 len always <=4"));
            i += u32::try_from(c.len_utf16()).expect("utf16 len always <=2");
        }
    }

    utf8_code_unit_offset
}

fn offset_to_source_location(
    offset: TextSize,
    text: &str,
    index: &LineIndex,
    encoding: PositionEncoding,
) -> SourceLocation {
    match encoding {
        PositionEncoding::UTF8 => {
            let row = index.line_index(offset);
            let column = offset - index.line_start(row, text);

            SourceLocation {
                column: OneIndexed::from_zero_indexed(column.into()),
                row,
            }
        }
        PositionEncoding::UTF16 => {
            let row = index.line_index(offset);

            let column = if index.is_ascii() {
                (offset - index.line_start(row, text)).into()
            } else {
                let up_to_line = &text[TextRange::new(index.line_start(row, text), offset)];
                up_to_line.encode_utf16().count()
            };

            SourceLocation {
                column: OneIndexed::from_zero_indexed(column),
                row,
            }
        }
        PositionEncoding::UTF32 => index.source_location(offset, text),
    }
}

fn source_location_to_position(location: &SourceLocation) -> types::Position {
    types::Position {
        line: u32::try_from(location.row.to_zero_indexed()).expect("row usize fits in u32"),
        character: u32::try_from(location.column.to_zero_indexed())
            .expect("character usize fits in u32"),
    }
}
