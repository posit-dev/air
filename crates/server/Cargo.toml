[package]
name = "server"
version = "0.1.1"
publish = false
authors = { workspace = true }
edition = { workspace = true }
rust-version = { workspace = true }
homepage = { workspace = true }
repository = { workspace = true }
license = { workspace = true }

[lib]

[dependencies]
air_r_factory = { workspace = true }
air_r_formatter = { workspace = true }
air_r_parser = { workspace = true }
air_r_syntax = { workspace = true }
source_file = { workspace = true }
workspace = { workspace = true }

anyhow = { workspace = true }
biome_formatter = { workspace = true }
biome_rowan = { workspace = true }
biome_text_size = { workspace = true }
crossbeam = { workspace = true }
dissimilar = { workspace = true }
ignore = { workspace = true }
itertools = { workspace = true }
jod-thread = { workspace = true }
lsp-server = { workspace = true }
lsp-types = { workspace = true }
regex = { workspace = true }
rustc-hash = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
thiserror = { workspace = true }
tracing = { workspace = true }
tracing-subscriber = { workspace = true, features = ["ansi", "local-time"] }
tree-sitter = { workspace = true }
tree-sitter-r = { workspace = true }
url = { workspace = true }
uuid = { workspace = true, features = ["v4"] }

[dev-dependencies]
assert_matches = { workspace = true }
insta = { workspace = true }
server_test = { workspace = true }

[target.'cfg(target_vendor = "apple")'.dependencies]
libc = { workspace = true }

[build-dependencies]
cargo_metadata.workspace = true

[lints]
workspace = true
