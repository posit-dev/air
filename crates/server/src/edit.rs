//! Types and utilities for working with text, modifying source files, and `Ruff <-> LSP` type conversion.

mod text_diff;
mod text_document;
mod text_edit;

use lsp_types::{PositionEncodingKind, Url};
pub(crate) use text_document::TextDocument;
pub(crate) use text_document::{DocumentVersion, LanguageId};
pub(crate) use text_edit::{Indel, TextEdit};

/// A convenient enumeration for supported text encodings. Can be converted to [`lsp_types::PositionEncodingKind`].
// Please maintain the order from least to greatest priority for the derived `Ord` impl.
#[derive(Default, Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum PositionEncoding {
    /// UTF 16 is the encoding supported by all LSP clients.
    #[default]
    UTF16,

    /// Second choice because UTF32 uses a fixed 4 byte encoding for each character (makes conversion relatively easy)
    UTF32,

    /// Ruff's preferred encoding
    UTF8,
}

/// A unique document ID, derived from a URL passed as part of an LSP request.
/// This document ID can point to either be a standalone Python file, a full notebook, or a cell within a notebook.
#[derive(Clone, Debug)]
pub enum DocumentKey {
    Text(Url),
}

impl DocumentKey {
    /// Converts the key back into its original URL.
    #[allow(dead_code)]
    pub(crate) fn into_url(self) -> Url {
        match self {
            DocumentKey::Text(url) => url,
        }
    }
}

impl std::fmt::Display for DocumentKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Text(url) => url.fmt(f),
        }
    }
}

impl From<PositionEncoding> for lsp_types::PositionEncodingKind {
    fn from(value: PositionEncoding) -> Self {
        match value {
            PositionEncoding::UTF8 => lsp_types::PositionEncodingKind::UTF8,
            PositionEncoding::UTF16 => lsp_types::PositionEncodingKind::UTF16,
            PositionEncoding::UTF32 => lsp_types::PositionEncodingKind::UTF32,
        }
    }
}

impl TryFrom<&lsp_types::PositionEncodingKind> for PositionEncoding {
    type Error = ();

    fn try_from(value: &PositionEncodingKind) -> Result<Self, Self::Error> {
        Ok(if value == &PositionEncodingKind::UTF8 {
            PositionEncoding::UTF8
        } else if value == &PositionEncodingKind::UTF16 {
            PositionEncoding::UTF16
        } else if value == &PositionEncodingKind::UTF32 {
            PositionEncoding::UTF32
        } else {
            return Err(());
        })
    }
}
