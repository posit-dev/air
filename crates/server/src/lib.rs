//! ## The Air Language Server

pub use server::Server;

#[macro_use]
mod message;

mod crates;
mod document;
mod error;
mod logging;
mod proto;
mod server;
mod session;

#[cfg(test)]
mod test;

pub(crate) const SERVER_NAME: &str = "Air Language Server";
pub(crate) const SERVER_VERSION: &str = env!("CARGO_PKG_VERSION");
