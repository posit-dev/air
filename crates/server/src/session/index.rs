use std::borrow::Cow;
use std::path::PathBuf;
use std::{path::Path, sync::Arc};

use lsp_types::Url;
use lsp_types::WorkspaceFolder;
use rustc_hash::FxHashMap;

use workspace::settings::Settings;

use crate::document::{DocumentKey, DocumentVersion, PositionEncoding, TextDocument};
use crate::session::workspaces::WorkspaceSettingsResolver;

/// Stores and tracks all open documents in a session, along with their associated settings.
#[derive(Default)]
pub(crate) struct Index {
    /// Maps all document file URLs to the associated document controller
    documents: FxHashMap<Url, DocumentController>,

    /// Maps a workspace folder root to its settings.
    settings: WorkspaceSettingsResolver,
}

/// A mutable handler to an underlying document.
#[derive(Debug)]
enum DocumentController {
    Text(Arc<TextDocument>),
}

/// A read-only query to an open document.
/// This query can 'select' a text document, but eventually could gain support for
/// selecting notebooks or individual notebook cells.
/// It also includes document settings.
#[derive(Clone)]
pub enum DocumentQuery {
    Text {
        file_url: Url,
        document: Arc<TextDocument>,
        settings: Arc<Settings>,
    },
}

impl Index {
    pub(super) fn new(workspace_folders: Vec<WorkspaceFolder>) -> anyhow::Result<Self> {
        Ok(Self {
            documents: FxHashMap::default(),
            settings: WorkspaceSettingsResolver::from_workspace_folders(workspace_folders),
        })
    }

    pub(super) fn text_document_urls(&self) -> impl Iterator<Item = &Url> + '_ {
        self.documents
            .iter()
            .filter(|(_, doc)| doc.as_text().is_some())
            .map(|(url, _)| url)
    }

    pub(super) fn update_text_document(
        &mut self,
        key: &DocumentKey,
        content_changes: Vec<lsp_types::TextDocumentContentChangeEvent>,
        new_version: DocumentVersion,
        encoding: PositionEncoding,
    ) -> anyhow::Result<()> {
        let controller = self.document_controller_for_key(key)?;
        let Some(document) = controller.as_text_mut() else {
            anyhow::bail!("Text document URI does not point to a text document");
        };
        document.apply_changes(content_changes, new_version, encoding);
        Ok(())
    }

    pub(super) fn key_from_url(&self, url: Url) -> DocumentKey {
        DocumentKey::Text(url)
    }

    pub(super) fn open_workspace_folder(&mut self, url: &Url) {
        self.settings.open_workspace_folder(url)
    }

    pub(super) fn close_workspace_folder(&mut self, url: &Url) {
        self.settings.close_workspace_folder(url)
    }

    pub(super) fn num_documents(&self) -> usize {
        self.documents.len()
    }

    pub(super) fn num_workspaces(&self) -> usize {
        self.settings.len()
    }

    pub(super) fn make_document_ref(&self, key: DocumentKey) -> Option<DocumentQuery> {
        let url = self.url_for_key(&key)?.clone();

        let settings = self.settings_for_url(&url);

        let controller = self.documents.get(&url)?;
        Some(controller.make_ref(url, settings))
    }

    /// Reloads relevant existing settings files based on a changed settings file path.
    pub(super) fn reload_settings(&mut self, changed_url: &Url) {
        self.settings.reload_workspaces_matched_by_url(changed_url);
    }

    pub(super) fn open_text_document(&mut self, url: Url, document: TextDocument) {
        self.documents
            .insert(url, DocumentController::new_text(document));
    }

    pub(super) fn close_document(&mut self, key: &DocumentKey) -> anyhow::Result<()> {
        let Some(url) = self.url_for_key(key).cloned() else {
            anyhow::bail!("Tried to close unavailable document `{key}`");
        };

        let Some(_) = self.documents.remove(&url) else {
            anyhow::bail!("tried to close document that didn't exist at {}", url)
        };
        Ok(())
    }

    // TODO: Index should manage per workspace client settings at some point once we have some
    // pub(super) fn client_settings(
    //     &self,
    //     key: &DocumentKey,
    //     global_settings: &ClientSettings,
    // ) -> ResolvedClientSettings {
    //     let Some(url) = self.url_for_key(key) else {
    //         return ResolvedClientSettings::global(global_settings);
    //     };
    //     let Some(WorkspaceSettings {
    //         client_settings, ..
    //     }) = self.settings_for_url(url)
    //     else {
    //         return ResolvedClientSettings::global(global_settings);
    //     };
    //     client_settings.clone()
    // }

    fn document_controller_for_key(
        &mut self,
        key: &DocumentKey,
    ) -> anyhow::Result<&mut DocumentController> {
        let Some(url) = self.url_for_key(key).cloned() else {
            anyhow::bail!("Tried to open unavailable document `{key}`");
        };
        let Some(controller) = self.documents.get_mut(&url) else {
            anyhow::bail!("Document controller not available at `{}`", url);
        };
        Ok(controller)
    }

    fn url_for_key<'a>(&'a self, key: &'a DocumentKey) -> Option<&'a Url> {
        match key {
            DocumentKey::Text(path) => Some(path),
        }
    }

    fn settings_for_url(&self, url: &Url) -> Arc<Settings> {
        self.settings.settings_for_url(url)
    }
}

impl DocumentController {
    fn new_text(document: TextDocument) -> Self {
        Self::Text(Arc::new(document))
    }

    fn make_ref(&self, file_url: Url, settings: Arc<Settings>) -> DocumentQuery {
        match &self {
            Self::Text(document) => DocumentQuery::Text {
                file_url,
                document: document.clone(),
                settings,
            },
        }
    }

    pub(crate) fn as_text(&self) -> Option<&TextDocument> {
        match self {
            Self::Text(document) => Some(document),
        }
    }

    pub(crate) fn as_text_mut(&mut self) -> Option<&mut TextDocument> {
        Some(match self {
            Self::Text(document) => Arc::make_mut(document),
        })
    }
}

impl DocumentQuery {
    /// Get the document settings associated with this query.
    pub(crate) fn settings(&self) -> &Settings {
        // Note that `&Arc<Settings>` nicely derefs to `&Settings` here automatically
        match self {
            Self::Text { settings, .. } => settings,
        }
    }

    /// Get the version of document selected by this query.
    #[allow(dead_code)]
    pub(crate) fn version(&self) -> DocumentVersion {
        match self {
            Self::Text { document, .. } => document.version(),
        }
    }

    /// Get the URL for the document selected by this query.
    pub(crate) fn file_url(&self) -> &Url {
        match self {
            Self::Text { file_url, .. } => file_url,
        }
    }

    /// Get the path for the document selected by this query.
    ///
    /// Returns `None` if this is an unsaved (untitled) document.
    ///
    /// The path isn't guaranteed to point to a real path on the filesystem. This is the case
    /// for unsaved (untitled) documents.
    #[allow(dead_code)]
    pub(crate) fn file_path(&self) -> Option<PathBuf> {
        self.file_url().to_file_path().ok()
    }

    /// Get the path for the document selected by this query, ignoring whether the file exists on disk.
    ///
    /// Returns the URL's path if this is an unsaved (untitled) document.
    #[allow(dead_code)]
    pub(crate) fn virtual_file_path(&self) -> Cow<Path> {
        self.file_path().map_or_else(
            || Cow::Borrowed(Path::new(self.file_url().path())),
            Cow::Owned,
        )
    }

    /// Attempt to access the single inner text document selected by the query.
    pub(crate) fn as_single_document(&self) -> &TextDocument {
        match self {
            Self::Text { document, .. } => document,
        }
    }
}
