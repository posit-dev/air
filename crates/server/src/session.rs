//! Data model, state management, and configuration resolution.

use std::sync::Arc;

use lsp_types::Url;
use lsp_types::WorkspaceFolder;

use crate::edit::{DocumentKey, DocumentVersion, PositionEncoding, TextDocument};

pub(crate) use self::capabilities::ResolvedClientCapabilities;
pub use self::index::DocumentQuery;

mod capabilities;
mod index;
mod workspaces;

/// The global state for the LSP
pub(crate) struct Session {
    /// Used to retrieve information about open documents and settings.
    index: index::Index,
    /// The global position encoding, negotiated during LSP initialization.
    position_encoding: PositionEncoding,
    /// Tracks what LSP features the client supports and doesn't support.
    resolved_client_capabilities: Arc<ResolvedClientCapabilities>,
}

/// An immutable snapshot of `Session` that references
/// a specific document.
pub(crate) struct DocumentSnapshot {
    #[allow(dead_code)]
    resolved_client_capabilities: Arc<ResolvedClientCapabilities>,
    document_ref: index::DocumentQuery,
    position_encoding: PositionEncoding,
}

impl Session {
    pub(crate) fn new(
        resolved_client_capabilities: ResolvedClientCapabilities,
        position_encoding: PositionEncoding,
        workspace_folders: Vec<WorkspaceFolder>,
    ) -> crate::Result<Self> {
        Ok(Self {
            position_encoding,
            index: index::Index::new(workspace_folders)?,
            resolved_client_capabilities: Arc::new(resolved_client_capabilities),
        })
    }

    pub(crate) fn key_from_url(&self, url: Url) -> DocumentKey {
        self.index.key_from_url(url)
    }

    /// Creates a document snapshot with the URL referencing the document to snapshot.
    pub(crate) fn take_snapshot(&self, url: Url) -> Option<DocumentSnapshot> {
        let key = self.key_from_url(url);
        Some(DocumentSnapshot {
            resolved_client_capabilities: self.resolved_client_capabilities.clone(),
            document_ref: self.index.make_document_ref(key)?,
            position_encoding: self.position_encoding,
        })
    }

    /// Iterates over the LSP URLs for all open text documents. These URLs are valid file paths.
    #[allow(dead_code)]
    pub(crate) fn text_document_urls(&self) -> impl Iterator<Item = &lsp_types::Url> + '_ {
        self.index.text_document_urls()
    }

    /// Updates a text document at the associated `key`.
    ///
    /// The document key must point to a text document, or this will throw an error.
    pub(crate) fn update_text_document(
        &mut self,
        key: &DocumentKey,
        content_changes: Vec<lsp_types::TextDocumentContentChangeEvent>,
        new_version: DocumentVersion,
    ) -> crate::Result<()> {
        let encoding = self.encoding();

        self.index
            .update_text_document(key, content_changes, new_version, encoding)
    }

    /// Registers a text document at the provided `url`.
    /// If a document is already open here, it will be overwritten.
    pub(crate) fn open_text_document(&mut self, url: Url, document: TextDocument) {
        self.index.open_text_document(url, document);
    }

    /// De-registers a document, specified by its key.
    /// Calling this multiple times for the same document is a logic error.
    pub(crate) fn close_document(&mut self, key: &DocumentKey) -> crate::Result<()> {
        self.index.close_document(key)?;
        Ok(())
    }

    /// Reloads the settings index
    pub(crate) fn reload_settings(&mut self, changed_url: &Url) {
        self.index.reload_settings(changed_url);
    }

    /// Open a workspace folder at the given `url`.
    pub(crate) fn open_workspace_folder(&mut self, url: &Url) -> crate::Result<()> {
        self.index.open_workspace_folder(url)
    }

    /// Close a workspace folder at the given `url`.
    pub(crate) fn close_workspace_folder(&mut self, url: &Url) -> crate::Result<()> {
        self.index.close_workspace_folder(url)?;
        Ok(())
    }

    #[allow(dead_code)]
    pub(crate) fn num_documents(&self) -> usize {
        self.index.num_documents()
    }

    #[allow(dead_code)]
    pub(crate) fn num_workspaces(&self) -> usize {
        self.index.num_workspaces()
    }

    #[allow(dead_code)]
    pub(crate) fn resolved_client_capabilities(&self) -> &ResolvedClientCapabilities {
        &self.resolved_client_capabilities
    }

    pub(crate) fn encoding(&self) -> PositionEncoding {
        self.position_encoding
    }
}

impl DocumentSnapshot {
    #[allow(dead_code)]
    pub(crate) fn resolved_client_capabilities(&self) -> &ResolvedClientCapabilities {
        &self.resolved_client_capabilities
    }

    pub fn query(&self) -> &index::DocumentQuery {
        &self.document_ref
    }

    pub(crate) fn encoding(&self) -> PositionEncoding {
        self.position_encoding
    }
}
