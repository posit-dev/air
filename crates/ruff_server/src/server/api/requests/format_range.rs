// +------------------------------------------------------------+
// | Code adopted from:                                         |
// | Repository: https://github.com/astral-sh/ruff.git          |
// | Commit: 5bc9d6d3aa694ab13f38dd5cf91b713fd3844380           |
// +------------------------------------------------------------+

use lsp_types::{self as types, request as req, Range};

use crate::edit::{PositionEncoding, TextDocument};
use crate::server::{client::Notifier, Result};
use crate::session::{DocumentQuery, DocumentSnapshot};

pub(crate) struct FormatRange;

impl super::RequestHandler for FormatRange {
    type RequestType = req::RangeFormatting;
}

impl super::BackgroundDocumentRequestHandler for FormatRange {
    super::define_document_url!(params: &types::DocumentRangeFormattingParams);
    fn run_with_snapshot(
        snapshot: DocumentSnapshot,
        _notifier: Notifier,
        params: types::DocumentRangeFormattingParams,
    ) -> Result<super::FormatResponse> {
        format_document_range(&snapshot, params.range)
    }
}

/// Formats the specified [`Range`] in the [`DocumentSnapshot`].
fn format_document_range(
    snapshot: &DocumentSnapshot,
    range: Range,
) -> Result<super::FormatResponse> {
    let text_document = snapshot.query().as_single_document();
    let query = snapshot.query();
    format_text_document_range(text_document, range, query, snapshot.encoding())
}

/// Formats the specified [`Range`] in the [`TextDocument`].
fn format_text_document_range(
    _text_document: &TextDocument,
    _range: Range,
    _query: &DocumentQuery,
    _encoding: PositionEncoding,
) -> Result<super::FormatResponse> {
    Ok(None)
    //     let document_settings = query.settings();
    //     let formatter_settings = &document_settings.format;
    //
    //     let text = text_document.contents();
    //     let index = text_document.index();
    //     let range = range.to_text_range(text, index, encoding);
    //
    //     let formatted_range = crate::format::format_range(
    //         text_document,
    //         query.source_type(),
    //         formatter_settings,
    //         range,
    //     )
    //     .with_failure_code(lsp_server::ErrorCode::InternalError)?;
    //
    //     Ok(formatted_range.map(|formatted_range| {
    //         vec![types::TextEdit {
    //             range: formatted_range
    //                 .source_range()
    //                 .to_range(text, index, encoding),
    //             new_text: formatted_range.into_code(),
    //         }]
    //     }))
}
