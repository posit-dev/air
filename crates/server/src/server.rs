//! Scheduling, I/O, and API endpoints.

use lsp_server as lsp;
use lsp_types as types;
use lsp_types::InitializeParams;
use std::num::NonZeroUsize;
// The new PanicInfoHook name requires MSRV >= 1.82
#[allow(deprecated)]
use std::panic::PanicInfo;
use types::DidChangeWatchedFilesRegistrationOptions;
use types::FileSystemWatcher;
use types::OneOf;
use types::TextDocumentSyncCapability;
use types::TextDocumentSyncKind;
use types::TextDocumentSyncOptions;
use types::WorkspaceFoldersServerCapabilities;

use self::connection::Connection;
use self::schedule::event_loop_thread;
use self::schedule::Scheduler;
use self::schedule::Task;
use crate::document::PositionEncoding;
use crate::message::try_show_message;
use crate::server::connection::ConnectionInitializer;
use crate::session::ResolvedClientCapabilities;
use crate::session::Session;

mod api;
mod client;
mod connection;
mod schedule;

pub(crate) use connection::ClientSender;

pub(crate) type Result<T> = std::result::Result<T, api::Error>;

pub struct Server {
    connection: Connection,
    client_capabilities: ResolvedClientCapabilities,
    worker_threads: NonZeroUsize,
    session: Session,
}

impl Server {
    pub fn new(
        worker_threads: NonZeroUsize,
        connection: lsp::Connection,
        connection_threads: Option<lsp::IoThreads>,
    ) -> anyhow::Result<Self> {
        let initializer = ConnectionInitializer::new(connection, connection_threads);

        let (id, initialize_params) = initializer.initialize_start()?;

        let client_capabilities = initialize_params.capabilities;
        let client_capabilities = ResolvedClientCapabilities::new(client_capabilities);
        let position_encoding = Self::find_best_position_encoding(&client_capabilities);
        let server_capabilities = Self::server_capabilities(position_encoding);

        let connection = initializer.initialize_finish(
            id,
            &server_capabilities,
            crate::SERVER_NAME,
            crate::SERVER_VERSION,
        )?;

        let InitializeParams {
            workspace_folders,
            client_info,
            ..
        } = initialize_params;

        let workspace_folders = workspace_folders.unwrap_or_default();

        // TODO: Get user specified options from `initialization_options`
        let log_level = None;
        let dependency_log_levels = None;

        crate::logging::init_logging(
            connection.make_sender(),
            log_level,
            dependency_log_levels,
            client_info,
        );

        crate::message::init_messenger(connection.make_sender());

        Ok(Self {
            connection,
            worker_threads,
            session: Session::new(
                client_capabilities.clone(),
                position_encoding,
                workspace_folders,
            )?,
            client_capabilities,
        })
    }

    pub fn run(self) -> anyhow::Result<()> {
        // The new PanicInfoHook name requires MSRV >= 1.82
        #[allow(deprecated)]
        type PanicHook = Box<dyn Fn(&PanicInfo<'_>) + 'static + Sync + Send>;
        struct RestorePanicHook {
            hook: Option<PanicHook>,
        }

        impl Drop for RestorePanicHook {
            fn drop(&mut self) {
                if let Some(hook) = self.hook.take() {
                    std::panic::set_hook(hook);
                }
            }
        }

        // unregister any previously registered panic hook
        // The hook will be restored when this function exits.
        let _ = RestorePanicHook {
            hook: Some(std::panic::take_hook()),
        };

        // When we panic, try to notify the client.
        std::panic::set_hook(Box::new(move |panic_info| {
            use std::io::Write;

            let backtrace = std::backtrace::Backtrace::force_capture();
            tracing::error!("{panic_info}\n{backtrace}");

            // we also need to print to stderr directly for when using `window/logMessage` because
            // the message won't be sent to the client.
            // But don't use `eprintln` because `eprintln` itself may panic if the pipe is broken.
            let mut stderr = std::io::stderr().lock();
            writeln!(stderr, "{panic_info}\n{backtrace}").ok();

            try_show_message(
                "The Air language server exited with a panic. See the logs for more details."
                    .to_string(),
                lsp_types::MessageType::ERROR,
            )
            .ok();
        }));

        event_loop_thread(move || {
            Self::event_loop(
                &self.connection,
                &self.client_capabilities,
                self.session,
                self.worker_threads,
            )?;
            self.connection.close()?;
            Ok(())
        })?
        .join()
    }

    #[allow(clippy::needless_pass_by_value)] // this is because we aren't using `next_request_id` yet.
    fn event_loop(
        connection: &Connection,
        resolved_client_capabilities: &ResolvedClientCapabilities,
        mut session: Session,
        worker_threads: NonZeroUsize,
    ) -> anyhow::Result<()> {
        let mut scheduler =
            schedule::Scheduler::new(&mut session, worker_threads, connection.make_sender());

        Self::try_register_capabilities(resolved_client_capabilities, &mut scheduler);
        for msg in connection.incoming() {
            if connection.handle_shutdown(&msg)? {
                break;
            }
            let task = match msg {
                lsp::Message::Request(req) => api::request(req),
                lsp::Message::Notification(notification) => api::notification(notification),
                lsp::Message::Response(response) => scheduler.response(response),
            };
            scheduler.dispatch(task);
        }

        Ok(())
    }

    fn try_register_capabilities(
        resolved_client_capabilities: &ResolvedClientCapabilities,
        scheduler: &mut Scheduler,
    ) {
        let _span = tracing::info_span!("try_register_capabilities").entered();

        // Register capabilities to the client
        let mut registrations: Vec<lsp_types::Registration> = vec![];

        if resolved_client_capabilities.dynamic_registration_for_did_change_watched_files {
            // Watch for changes in `air.toml` files so we can react dynamically
            let watch_air_toml_registration = lsp_types::Registration {
                id: String::from("air-toml-watcher"),
                method: "workspace/didChangeWatchedFiles".into(),
                register_options: Some(
                    serde_json::to_value(DidChangeWatchedFilesRegistrationOptions {
                        watchers: vec![FileSystemWatcher {
                            glob_pattern: lsp_types::GlobPattern::String("**/air.toml".into()),
                            kind: None,
                        }],
                    })
                    .unwrap(),
                ),
            };

            registrations.push(watch_air_toml_registration);
        } else {
            tracing::warn!("LSP client does not support watched files dynamic capability - automatic configuration reloading will not be available.");
        }

        if !registrations.is_empty() {
            let params = lsp_types::RegistrationParams { registrations };

            let response_handler = |()| {
                tracing::info!("Dynamic configuration successfully registered");
                Task::nothing()
            };

            if let Err(error) = scheduler
                .request::<lsp_types::request::RegisterCapability>(params, response_handler)
            {
                tracing::error!(
                    "An error occurred when trying to dynamically register capabilities: {error}"
                );
            }
        }
    }

    fn find_best_position_encoding(
        client_capabilities: &ResolvedClientCapabilities,
    ) -> PositionEncoding {
        // If the client supports UTF-8 we use that, even if it's not its
        // preferred encoding (at position 0). Otherwise we use the mandatory
        // UTF-16 encoding that all clients and servers must support, even if
        // the client would have preferred UTF-32. Note that VSCode and Positron
        // only support UTF-16.
        if client_capabilities
            .position_encodings
            .contains(&lsp_types::PositionEncodingKind::UTF8)
        {
            PositionEncoding::UTF8
        } else {
            PositionEncoding::UTF16
        }
    }

    fn server_capabilities(position_encoding: PositionEncoding) -> types::ServerCapabilities {
        types::ServerCapabilities {
            position_encoding: Some(position_encoding.into()),
            text_document_sync: Some(TextDocumentSyncCapability::Options(
                TextDocumentSyncOptions {
                    open_close: Some(true),
                    change: Some(TextDocumentSyncKind::INCREMENTAL),
                    will_save: Some(false),
                    will_save_wait_until: Some(false),
                    ..Default::default()
                },
            )),
            workspace: Some(types::WorkspaceServerCapabilities {
                workspace_folders: Some(WorkspaceFoldersServerCapabilities {
                    supported: Some(true),
                    change_notifications: Some(OneOf::Left(true)),
                }),
                file_operations: None,
            }),
            document_formatting_provider: Some(OneOf::Left(true)),
            document_range_formatting_provider: Some(OneOf::Left(true)),
            ..Default::default()
        }
    }
}
