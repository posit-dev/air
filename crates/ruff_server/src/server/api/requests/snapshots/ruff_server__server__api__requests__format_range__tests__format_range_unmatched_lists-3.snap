---
source: crates/ruff_server/src/server/api/requests/format_range.rs
expression: output3
---
0+0
1 + 1
{
	2 + 2
}
3 + 3
